### DISCUSSION TITLE
Advice on using mice package for imputation with propensity score matching on a hierarchical dataset

### DISCUSSION THREAD

**Discussion by ayoubmakboul** [amices/mice#561] [Impute, analyse and pool]
Hello,

I am currently working on analyzing the PIRLS dataset, which has a hierarchical structure and a large dimension. I am dealing with missing data in this dataset and I plan to incorporate propensity score matching (PSM) into my analysis.

My specific request for advice revolves around using the mice package for imputation in this context. Here are some details:

Dataset: PIRLS dataset
Hierarchical structure: The dataset has a hierarchical structure with nested levels (e.g., students within schools or classes).
Missing data: The dataset contains missing values that need to be imputed before performing PSM.
Missing data mechanism: I have investigated the missing data mechanism and assume that the variables of interest are missing at random (MAR). While I have a clear idea about the predictor variables for the imputation model. 
I want to exclude some variables from imputation but I want to keep them on the database ( the weight of the sample, the ID of the student, and the ID of the class ... ) 
how I can use the weight in the imputation process) ?

**Comment by gerkovink** [comment]
I suggest you have a look at the [`mice` vignettes](https://www.gerkovink.com/miceVignettes/). Together, these vignettes cover all skills related to your question.

**Reply by ayoubmakboul** [reply]
I am delighted to receive your prompt response and appreciate your willingness to assist me with my request.
I have already reviewed the vignettes and your article titled "Partitioned predictive mean matching as a multilevel imputation technique" which was published in 2015. However, I am still facing issues with the code, particularly in manipulating the prediction matrix. I tried exporting it to Excel as mentioned in the vignettes, but I am unable to import and use it as a predictor. Even when I attempted to convert it back to a matrix (using as.matrix), I encountered challenges.

**Comment by gerkovink** [comment]
Can you provide a [reprex](https://reprex.tidyverse.org/)? It's really challenging to troubleshoot, otherwise.

**Reply by ayoubmakboul** [reply]
"Thank you! I found the solution, and it was surprisingly simple. Here's what I did: 

require(openxlsx)

fmodified_file_path <- "C:/work/imputation process/imputation/pdm.xlsx"
modified_pred_df <- read.xlsx(modified_file_path)
modified_pred <- as.matrix(modified_pred_df)


Get the column and row names from the 'pred' matrix
col_names <- colnames(pred)
row_names <- rownames(pred)


Create the 'pred1' matrix using Book1.R and assign the column and row names
pred1 <- as.matrix(Book1)
colnames(pred1) <- col_names
rownames(pred1) <- row_names
###################
I have a question regarding the 'mice' function. Can I simply add the argument for simple weights, or is there something specific I need to consider?

for example   :    >R  weights <- "weights_variable" 
                           >     imp <- mice ( data = data ,m= 5 , pred = pred , meth = meth , weights = weights.... )
